File: backtracking/22.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 0:
            return []
        ret = []
        self.dfs(ret, '(', 1, 0, n)
        return ret
    
    def dfs(self, ret, temp, left, right, n):
        if len(temp) == n*2 - 1:
            if left - right == 1:
                ret.append(temp+')')
                return
        else:
            if left < right:
                return
            else:
                self.dfs(ret, temp+'(', left+1, right, n)
                self.dfs(ret, temp+')', left, right+1, n)
        
    

File: backtracking/22.cc
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        recur("", n , 0);
        return ret;
    }
private:
    void recur(std::string cur, int open, int close) {
        if (open == 0 && close == 0) {
            ret.emplace_back(cur);
            return;
        }
        if (open > 0) {
            recur(cur + "(", open-1, close+1);
        }
        if (close > 0)
            recur(cur + ")", open, close-1);
    }
    std::vector<string> ret;
};
